nameresolution/consul/consul.go

```diff
@@ -0,0 +1,273 @@
+package consul
+
+import (
+	"crypto/rand"
+	"fmt"
+	"math/big"
+	"strconv"
+
+	consul "github.com/hashicorp/consul/api"
+
+	nr "github.com/dapr/components-contrib/nameresolution"
+	"github.com/dapr/dapr/pkg/logger"
+)
+
+const daprMeta string = "DAPR_PORT" // default key for DAPR_PORT metadata
+
+type client struct {
+	*consul.Client
+}
+
+func (c *client) InitClient(config *consul.Config) error {
+	var err error
+
+	c.Client, err = consul.NewClient(config)
+	if err != nil {
+		return fmt.Errorf("consul api error initing client: %w", err)
+	}
+
+	return nil
+}
+
+func (c *client) Agent() agentInterface {
+	return c.Client.Agent()
+}
+
+func (c *client) Health() healthInterface {
+	return c.Client.Health()
+}
+
+type clientInterface interface {
+	InitClient(config *consul.Config) error
+	Agent() agentInterface
+	Health() healthInterface
+}
+
+type agentInterface interface {
+	Self() (map[string]map[string]interface{}, error)
+	ServiceRegister(service *consul.AgentServiceRegistration) error
+}
+
+type healthInterface interface {
+	Service(service, tag string, passingOnly bool, q *consul.QueryOptions) ([]*consul.ServiceEntry, *consul.QueryMeta, error)
+}
+
+type resolver struct {
+	config resolverConfig
+	logger logger.Logger
+	client clientInterface
+}
+
+type resolverConfig struct {
+	Client          *consul.Config
+	QueryOptions    *consul.QueryOptions
+	Registration    *consul.AgentServiceRegistration
+	DaprPortMetaKey string
+}
+
+// NewResolver creates Consul name resolver.
+func NewResolver(logger logger.Logger) nr.Resolver {
+	return newResolver(logger, resolverConfig{}, &client{})
+}
+
+func newResolver(logger logger.Logger, resolverConfig resolverConfig, client clientInterface) nr.Resolver {
+	return &resolver{
+		logger: logger,
+		config: resolverConfig,
+		client: client,
+	}
+}
+
+// Init will configure component. It will also register service or validate client connection based on config
+func (r *resolver) Init(metadata nr.Metadata) error {
+	var err error
+
+	r.config, err = getConfig(metadata)
+	if err != nil {
+		return err
+	}
+
+	if err = r.client.InitClient(r.config.Client); err != nil {
+		return fmt.Errorf("failed to init consul client: %w", err)
+	}
+
+	// register service to consul
+	if r.config.Registration != nil {
+		if err := r.client.Agent().ServiceRegister(r.config.Registration); err != nil {
+			return fmt.Errorf("failed to register consul service: %w", err)
+		}
+
+		r.logger.Infof("service:%s registered on consul agent", r.config.Registration.Name)
+	} else if _, err := r.client.Agent().Self(); err != nil {
+		return fmt.Errorf("failed check on consul agent: %w", err)
+	}
+
+	return nil
+}
+
+// ResolveID resolves name to address via consul
+func (r *resolver) ResolveID(req nr.ResolveRequest) (string, error) {
+	cfg := r.config
+	services, _, err := r.client.Health().Service(req.ID, "", true, cfg.QueryOptions)
```

Comment:
What is the performance implication of calling this each time? Is the Consul API invoked on every request? I'm hoping there is some sort of caching involved. My years ago, I implemented service discovery w/ Consul using a Watch to keep everything in sync in memory. I'm not sure if that is a better approach but I at least wanted to understand how this would perform at scale.

Reply:
Yes, it will make a call to the local agent every time and that is where the caching sits. I've not done any benchmarking but a simple loop with 10k requests gives me an average of 1ms per round trip. No doubt, if we implement the watcher then that latency should shrink but I'll need a bit more time to get to that

Reply:
I think it's fine to leave this as is now and iterate on that with caching later on.

Reply:
I agree. Since it's to the local agent it's not that bad.

Reply:
@pkedy @yaron2 just an fyi I'm currently brewing this one and it's in testing, I'll PR it separately once it's ready